# Cleaning the data

In [2]:
import pandas as pd
import numpy as np

Read in ratings and movies data

In [3]:
ratings_messy = pd.read_csv("IMDb ratings.csv")
films_messy = pd.read_csv("IMDb movies.csv")

/opt/anaconda3/lib/python3.8/site-packages/IPython/core/interactiveshell.py:3441: DtypeWarning: Columns (3) have mixed types.Specify dtype option on import or set low_memory=False.
  exec(code_obj, self.user_global_ns, self.user_ns)


Set title IDs as index to smoothly join both tables, then drop unnecessary columns.

In [4]:
films_messy = films_messy.set_index('imdb_title_id')
ratings_messy = ratings_messy.set_index('imdb_title_id')

Check how many movies we can work with in this data, since metascore is our response variable.

In [13]:
len(films_messy['metascore'].dropna())

13305

Join both tables, then drop unnecessary columns.

In [9]:
combined_messy = pd.concat([films_messy, ratings_messy], axis=1)
combined_messy.columns

Index(['title', 'original_title', 'year', 'date_published', 'genre',
       'duration', 'country', 'language', 'director', 'writer',
       'production_company', 'actors', 'description', 'avg_vote', 'votes',
       'budget', 'usa_gross_income', 'worlwide_gross_income', 'metascore',
       'reviews_from_users', 'reviews_from_critics', 'weighted_average_vote',
       'total_votes', 'mean_vote', 'median_vote', 'votes_10', 'votes_9',
       'votes_8', 'votes_7', 'votes_6', 'votes_5', 'votes_4', 'votes_3',
       'votes_2', 'votes_1', 'allgenders_0age_avg_vote',
       'allgenders_0age_votes', 'allgenders_18age_avg_vote',
       'allgenders_18age_votes', 'allgenders_30age_avg_vote',
       'allgenders_30age_votes', 'allgenders_45age_avg_vote',
       'allgenders_45age_votes', 'males_allages_avg_vote',
       'males_allages_votes', 'males_0age_avg_vote', 'males_0age_votes',
       'males_18age_avg_vote', 'males_18age_votes', 'males_30age_avg_vote',
       'males_30age_votes', 'males_45age_av

In [79]:
ratings = combined_messy.drop(['original_title',
                               'date_published',
                               'duration',
                               'country',
                               'language',
                               'production_company',
                               'description',
                               'avg_vote',
                               'votes',
                               'budget',
                               'usa_gross_income',
                               'worlwide_gross_income',
                               'reviews_from_users',
                               'reviews_from_critics',
                               'allgenders_0age_avg_vote', # Dropping this because so many NAs
                               'allgenders_0age_votes',
                               'allgenders_18age_votes',
                               'allgenders_30age_votes',
                               'allgenders_45age_votes',
                               'males_allages_votes',
                               'males_0age_avg_vote',      # Dropping this because so many NAs
                               'males_0age_votes',
                               'males_18age_votes',
                               'males_30age_votes',
                               'males_45age_votes',
                               'females_allages_votes',
                               'females_0age_avg_vote',    # Dropping this because so many NAs
                               'females_0age_votes',
                               'females_18age_votes',
                               'females_30age_votes',
                               'females_45age_votes',
                               'top1000_voters_rating',
                               'top1000_voters_votes',
                               'us_voters_rating',
                               'us_voters_votes',
                               'non_us_voters_rating',
                               'non_us_voters_votes'],
                              1)

/var/folders/5r/j3mtn4sx1wv_xsdwft49lcdm0000gn/T/ipykernel_76448/4007758708.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  ratings = combined_messy.drop(['original_title',


As alluded to in the commented code, we planned to keep all of the average vote columns, but including the ones from critics under the age of 18 left a lot of missing data. This is presumably because it's not very common for people younger than 18 to be professional movie critics. The only column we want to lose data to is the metascore column, as that is our response.

In [80]:
ratings = ratings.dropna()

In [81]:
len(ratings)

#len(films_messy['metascore'].dropna()) - len(ratings)

13193

This looks good. We're only missing 112 movies that have a reported metascore to missing values from other columns.

In [82]:
ratings.head()

title  year                  genre  \
imdb_title_id                                                     
tt0006864              Intolerance  1916         Drama, History   
tt0017136               Metropolis  1927          Drama, Sci-Fi   
tt0018037      Il cantante di jazz  1927  Drama, Music, Musical   
tt0018773                 Il circo  1928        Comedy, Romance   
tt0019777            The Cocoanuts  1929        Comedy, Musical   

                                    director  \
imdb_title_id                                  
tt0006864                      D.W. Griffith   
tt0017136                         Fritz Lang   
tt0018037                      Alan Crosland   
tt0018773                    Charles Chaplin   
tt0019777      Robert Florey, Joseph Santley   

                                          writer  \
imdb_title_id                                      
tt0006864              D.W. Griffith, Anita Loos   
tt0017136       Thea von Harbou, Thea von Harbou   
tt0018037      Samson Raphaelson, Alfred A. Cohn   
tt0018773                        Charles Chaplin   
tt0019777      George S. Kaufman, Morrie Ryskind   

                                                          actors  metascore  \
imdb_title_id                                                                 
tt0006864      Lillian Gish, Mae Marsh, Robert Harron, F.A. T...       99.0   
tt0017136      Alfred Abel, Gustav Fröhlich, Rudolf Klein-Rog...       98.0   
tt0018037      Al Jolson, May McAvoy, Warner Oland, Eugenie B...       66.0   
tt0018773      Al Ernest Garcia, Merna Kennedy, Harry Crocker...       90.0   
tt0019777      The Marx Brothers, Zeppo Marx, Groucho Marx, H...       69.0   

               weighted_average_vote  total_votes  mean_vote  ...  \
imdb_title_id                                                 ...   
tt0006864                        7.8        13875        7.8  ...   
tt0017136                        8.3       156076        8.1  ...   
tt0018037                        6.5         8866        6.8  ...   
tt0018773                        8.1        27414        7.9  ...   
tt0019777                        7.0         6900        7.2  ...   

               allgenders_30age_avg_vote  allgenders_45age_avg_vote  \
imdb_title_id                                                         
tt0006864                            7.7                        7.8   
tt0017136                            8.2                        8.3   
tt0018037                            6.4                        6.6   
tt0018773                            8.1                        8.1   
tt0019777                            6.9                        7.1   

               males_allages_avg_vote  males_18age_avg_vote  \
imdb_title_id                                                 
tt0006864                         7.8                   7.9   
tt0017136                         8.3                   8.3   
tt0018037                         6.4                   6.5   
tt0018773                         8.1                   8.2   
tt0019777                         6.9                   6.8   

               males_30age_avg_vote  males_45age_avg_vote  \
imdb_title_id                                               
tt0006864                       7.7                   7.8   
tt0017136                       8.2                   8.3   
tt0018037                       6.4                   6.6   
tt0018773                       8.1                   8.1   
tt0019777                       6.8                   7.1   

               females_allages_avg_vote  females_18age_avg_vote  \
imdb_title_id                                                     
tt0006864                           7.5                     7.2   
tt0017136                           8.3                     8.2   
tt0018037                           6.6                     6.7   
tt0018773                           8.2                     8.3   
tt0019777                           7.2          

# Integrate genres into the clean dataframe

There are 13193 films in our dataset, each of which have one or more genres associated with them in the following format:  
  
tt0006864              Drama, History  
tt0017136               Drama, Sci-Fi  
tt0018037       Drama, Music, Musical  
tt0018773             Comedy, Romance  
tt0019777             Comedy, Musical  
                       ...             
tt9845398                       Drama  
tt9883996    Biography, Comedy, Drama  
tt9892094                       Drama  
tt9894470       Action, Crime, Horror  
tt9898858              Action, Comedy  
  
We see that some films have multiple genres in the same string, so we have some cleaning to do. We need to figure out a way to find all the unique genres and create columns for each one, indicating which genre is associated with a given film using a 1 (True) or 0 (False).

Start by creating a list of the genres for each movie.

In [20]:
genres = ratings.genre

The following code iterates through the list of movie's genres and, if there are multiple, splits them into lists. If we come across a genre that we haven't seen yet, we add it to the list ```genres_unique```.

In [83]:
genres_unique = []

for film in range(0,len(genres)):
    film_genres = genres[film].split(", ")
    for genre in film_genres:
        if (genre not in genres_unique):
            genres_unique.append(genre)

Let's check if the list is indeed unique:

In [84]:
#pd.Series(genres_unique).nunique()
# 21
#len(genres_unique)
# 21

There are 21 unique genres associated with the films in the dataset. Now we will cycle through each genre, creating a column for each one and assigning 1 if the film contains that genre, and 0 if it doesn't.

In [85]:
for genre in genres_unique:
    ratings[genre] = ratings['genre'].str.contains(genre).astype(int)

# Drop the messy genres column
ratings = ratings.drop('genre', axis=1)

The dataset now includes genre data in one-hot encoded form.

In [86]:
ratings.head()

title  year                       director  \
imdb_title_id                                                             
tt0006864              Intolerance  1916                  D.W. Griffith   
tt0017136               Metropolis  1927                     Fritz Lang   
tt0018037      Il cantante di jazz  1927                  Alan Crosland   
tt0018773                 Il circo  1928                Charles Chaplin   
tt0019777            The Cocoanuts  1929  Robert Florey, Joseph Santley   

                                          writer  \
imdb_title_id                                      
tt0006864              D.W. Griffith, Anita Loos   
tt0017136       Thea von Harbou, Thea von Harbou   
tt0018037      Samson Raphaelson, Alfred A. Cohn   
tt0018773                        Charles Chaplin   
tt0019777      George S. Kaufman, Morrie Ryskind   

                                                          actors  metascore  \
imdb_title_id                                                                 
tt0006864      Lillian Gish, Mae Marsh, Robert Harron, F.A. T...       99.0   
tt0017136      Alfred Abel, Gustav Fröhlich, Rudolf Klein-Rog...       98.0   
tt0018037      Al Jolson, May McAvoy, Warner Oland, Eugenie B...       66.0   
tt0018773      Al Ernest Garcia, Merna Kennedy, Harry Crocker...       90.0   
tt0019777      The Marx Brothers, Zeppo Marx, Groucho Marx, H...       69.0   

               weighted_average_vote  total_votes  mean_vote  median_vote  \
imdb_title_id                                                               
tt0006864                        7.8        13875        7.8          8.0   
tt0017136                        8.3       156076        8.1          8.0   
tt0018037                        6.5         8866        6.8          7.0   
tt0018773                        8.1        27414        7.9          8.0   
tt0019777                        7.0         6900        7.2          7.0   

               ...  Adventure  Family  Fantasy  Mystery  Thriller  Biography  \
imdb_title_id  ...                                                             
tt0006864      ...          0       0        0        0         0          0   
tt0017136      ...          0       0        0        0         0          0   
tt0018037      ...          0       0        0        0         0          0   
tt0018773      ...          0       0        0        0         0          0   
tt0019777      ...          0       0        0        0         0          0   

               Animation  Action  Film-Noir  Sport  
imdb_title_id                                       
tt0006864              0       0          0      0  
tt0017136              0       0          0      0  
tt0018037              0       0          0      0  
tt0018773              0       0          0      0  
tt0019777              0       0          0      0  

[5 rows x 52 columns]

# Integrate directors, writers, and actors into clean dataframe

In [37]:
ratings.director

imdb_title_id
tt0006864                    D.W. Griffith
tt0017136                       Fritz Lang
tt0018037                    Alan Crosland
tt0018773                  Charles Chaplin
tt0019777    Robert Florey, Joseph Santley
                         ...              
tt9845398                     Lucio Castro
tt9883996                        Euros Lyn
tt9892094                     Bassam Tariq
tt9894470                        Joe Begos
tt9898858                    Michael Dowse
Name: director, Length: 13193, dtype: object

The directors, writers, and actors, like the genres, have many different unique values and some with multiple people. For example, here are the directors:
  
tt0006864                    D.W. Griffith  
tt0017136                       Fritz Lang  
tt0018037                    Alan Crosland  
tt0018773                  Charles Chaplin  
tt0019777    Robert Florey, Joseph Santley  
                         ...                
tt9845398                     Lucio Castro  
tt9883996                        Euros Lyn  
tt9892094                     Bassam Tariq  
tt9894470                        Joe Begos  
tt9898858                    Michael Dowse  
  
For directors, writers, and actors, we need to perform a similar process as the process we performed above for the genres.

In [28]:
directors = ratings.director
writers = ratings.writer
actors = ratings.actors

The following code iterates through each list and, if there are multiple, splits them into lists. If we come across an actor/director/writer that we haven't seen yet, we add it to the respective list.

In [53]:
directors_unique = []

for film in range(0,len(directors)):
    film_directors = directors[film].split(", ")
    for director in film_directors:
        if (director not in directors_unique):
            directors_unique.append(director)

In [44]:
writers_unique = []

for film in range(0,len(writers)):
    film_writers = writers[film].split(", ")
    for writer in film_writers:
        if (writer not in writers_unique):
            writers_unique.append(writer)

In [48]:
actors_unique = []
         
for film in range(0,len(actors)):
    film_actors = actors[film].split(", ")
    for actor in film_actors:
        if (actor not in actors_unique):
            actors_unique.append(actor)

Let's check if the lists are indeed unique:

In [55]:
#pd.Series(directors_unique).nunique()
# 6573
#len(directors_unique)
# 6573

#pd.Series(writers_unique).nunique()
# 12317
#len(writers_unique)
# 12317

#pd.Series(actors_unique).nunique()
# 92772
#len(actors_unique)
# 92772


6573

For column-naming conventions, we will now cycle through each of these lists, replacing spaces with underscores and adding whether the person is a director, writer, or actor.

In [62]:
directors_clean = []

for director in directors_unique:
    director_underscore = director.replace(" ", "_")
    director_underscore += "_dir"
    directors_clean.append(director_underscore)

In [66]:
writers_clean = []

for writer in writers_unique:
    writer_underscore = writer.replace(" ", "_")
    writer_underscore += "_wri"
    writers_clean.append(writer_underscore)

In [68]:
actors_clean = []

for actor in actors_unique:
    actor_underscore = actor.replace(" ", "_")
    actor_underscore += "_act"
    actors_clean.append(actor_underscore)

There are 6573 unique directors, 12317 unique writers, and 92772 unique actors. Now we will cycle through each of these lists, creating a column for each element and assigning 1 if the film contains that director, writer, or actor, and 0 if it doesn't.

In [87]:
for director_ind in range(0,len(directors_clean)):
    ratings[directors_clean[director_ind]] = ratings['director'].str.contains(directors_unique[director_ind]).astype(int)

# Drop the messy genres column
ratings = ratings.drop('director', axis=1)

/var/folders/5r/j3mtn4sx1wv_xsdwft49lcdm0000gn/T/ipykernel_76448/1381502355.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  ratings[directors_clean[director_ind]] = ratings['director'].str.contains(directors_unique[director_ind]).astype(int)


In [88]:
ratings[['D.W._Griffith_dir']]

D.W._Griffith_dir
imdb_title_id                   
tt0006864                      1
tt0017136                      0
tt0018037                      0
tt0018773                      0
tt0019777                      0
...                          ...
tt9845398                      0
tt9883996                      0
tt9892094                      0
tt9894470                      0
tt9898858                      0

[13193 rows x 1 columns]

In [89]:
for writer_ind in range(0,len(writers_clean)):
    ratings[writers_clean[writer_ind]] = ratings['writer'].str.contains(writers_unique[writer_ind]).astype(int)

# Drop the messy genres column
ratings = ratings.drop('writer', axis=1)

/var/folders/5r/j3mtn4sx1wv_xsdwft49lcdm0000gn/T/ipykernel_76448/2574969597.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  ratings[writers_clean[writer_ind]] = ratings['writer'].str.contains(writers_unique[writer_ind]).astype(int)


In [91]:
for actor_ind in range(0,len(actors_clean)):
    ratings[actors_clean[actor_ind]] = ratings['actors'].str.contains(actors_unique[actor_ind]).astype(int)

# Drop the messy genres column
ratings = ratings.drop('actors', axis=1)

/var/folders/5r/j3mtn4sx1wv_xsdwft49lcdm0000gn/T/ipykernel_76448/1034067502.py:2: PerformanceWarning: DataFrame is highly fragmented.  This is usually the result of calling `frame.insert` many times, which has poor performance.  Consider joining all columns at once using pd.concat(axis=1) instead.  To get a de-fragmented frame, use `newframe = frame.copy()`
  ratings[actors_clean[actor_ind]] = ratings['actors'].str.contains(actors_unique[actor_ind]).astype(int)


We will now export the complete dataset so it is available for analysis

In [129]:
#ratings.to_csv('ratings_complete.csv')

Never mind, that was running for about an hour and a half and got up to 3 GB. We probably won't use that dataset anyways.

Now the dataset has thousands of director, writer, and actor columns. There's no purpose in having columns for directors, writers, and actors who only have a small number of works tied to their name in this position. With the complete dataset now on our hands, we will drop all directors, writers, and actors who have directed/written/acted in two films or fewer.

In [121]:
# Store the names of all the columns in question into a list
hotcoded_cols = ratings.columns[49:]

In [125]:
cols_to_drop = []

for col in hotcoded_cols:
    if (sum(ratings[col]) <= 2):
        cols_to_drop.append(col)

In [126]:
len(cols_to_drop)

93559

In [127]:
len(ratings.columns)

111711

In [130]:
ratings_refined = ratings.drop(cols_to_drop, axis=1)

Export data:

In [133]:
ratings_refined.to_csv('ratings_refined.csv')